modules/local-caching-classloader/src/main/java/org/apache/accumulo/classloader/lcc/state/ContextClassLoader.java

```diff
@@ -0,0 +1,230 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.accumulo.classloader.lcc.state;
+
+import java.io.File;
+import java.lang.ref.WeakReference;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.nio.channels.FileChannel;
+import java.nio.channels.FileLock;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Set;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+
+import org.apache.accumulo.classloader.lcc.cache.CacheUtils;
+import org.apache.accumulo.classloader.lcc.manifest.ContextDefinition;
+import org.apache.accumulo.classloader.lcc.manifest.Manifest;
+import org.apache.accumulo.classloader.lcc.manifest.Resource;
+import org.apache.accumulo.classloader.lcc.resolvers.FileResolver;
+import org.apache.accumulo.core.spi.common.ContextClassLoaderFactory.ContextClassLoaderException;
+import org.apache.accumulo.core.util.Pair;
+import org.apache.commons.codec.digest.DigestUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ContextClassLoader {
+
+  public static class ClassPathElement {
+    private final FileResolver remote;
+    private final URL localCachedCopy;
+    private final String localCachedCopyDigest;
+
+    public ClassPathElement(FileResolver remote, URL localCachedCopy,
+        String localCachedCopyDigest) {
+      this.remote = remote;
+      this.localCachedCopy = localCachedCopy;
+      this.localCachedCopyDigest = localCachedCopyDigest;
+    }
+
+    public FileResolver getRemote() {
+      return remote;
+    }
+
+    public URL getLocalCachedCopy() {
+      return localCachedCopy;
+    }
+
+    public String getLocalCachedCopyDigest() {
+      return localCachedCopyDigest;
+    }
+  }
+
+  private ClassPathElement cacheResource(final Resource resource) throws Exception {
+
+    final DigestUtils digest = new DigestUtils("MD5");
+    final FileResolver source = FileResolver.resolve(resource.getURL());
+    final Path cacheLocation =
+        contextCacheDir.resolve(source.getFileName() + "_" + resource.getChecksum());
+    final File cacheFile = cacheLocation.toFile();
+    if (!Files.exists(cacheLocation)) {
+      Files.copy(source.getInputStream(), cacheLocation);
+      String md5 = digest.digestAsHex(cacheFile);
+      if (!resource.getChecksum().equals(md5)) {
+        // TODO: What we just wrote does not match the Manifest.
```

Comment:
Maybe we throw an Exceptionhere since we'd have to repair the manifest at this point. Also this highlights that we're relying on our checksum computation here matching whatever is done in Resource.getChecksum(). Wondering if maybe it'd be good to centralize the checksum behind some static utility to enforce the consistent algorithm?